_posts/2020-10-26-dynamically-changeable-logging.md

```diff
@@ -0,0 +1,150 @@
+Topic Operator
+User Operator
+Strimzi Cluster Operator
+
+You can change logging levels dynamically by ...
+
+Each implementation requires a different approach, which we will look at now.
+
+
+### Strimzi Cluster Operator
+
+In order to change the Cluster Operator logging config, a simple change is required.
+You have to configure the logging settings in the `ConfigMap` used for Cluster Operator configuration.
+This `ConfigMap` is by default called `strimzi-cluster-operator`,
+To edit the logging configuration use
+```
+$ kubectl edit cm strimzi-cluster-operator
+```
+
+Under the field `log4j2.properties` you should see current logging configuration.
+Now you can edit this configuration to whatever you want with one exception.
+The entry `monitorInterval` must stay in the place.
+
+The second way to change `ConfigMap` is to modify the file `install/cluster-operator/050-ConfigMap-strimzi-cluster-operator.yaml`, and apply the changes by this command:
+
+```
+$ kubectl apply -f install/cluster-operator/050-ConfigMap-strimzi-cluster-operator.yaml
+```
+
+The same restrictions apply.
+The entry `monitorInterval` must stay in the place.
+
+The change in logging takes some time to be applied. 
+You can find out more in the implementation details section of this post.
+
+### Topic Operator, User Operator and Kafka resources
+
+The approach to changing the logging configuration of User Operator, Topic Operator, Kafka Bridge, Kafka brokers, Kafka Connect and Kafka MirrorMaker2 did not change.
+The logging is still configurable via `inline` or `external` logging.
+
+Example of `inline` logging.
+```yaml
+...
+spec:
+  logging:
+    type: inline
+      loggers:
+        kafka.root.logger.level = INFO
+...
+```
+
+Example of external logging.
+```yaml
+...
+spec:
+  logging:
+    type: external
+      name: my-external-config-map
+...
+```
+
+Only loggers can be changed dynamically.
+Logging appenders are changed using an external `ConfigMap` and the rolling update is triggered. This does not apply for Kafka Bridge, User operator and Topic operator.
+
+
+## Verifying logging levels are set correctly
+
+In most cases, it takes a few seconds to change the logging level, and then you can observe the change in the logs.
+If you want to be sure the loggers and logging levels are changed, you can use these commands.
+
+For Kafka brokers:
+
+```
+$ kubectl  exec <kafka-cluster-name>-zookeeper-0 -- bin/kafka-configs.sh --bootstrap-server <kafka-cluster-name>-kafka-bootstrap:9092 --entity-type broker-loggers --entity-name 0 --describe
+```
+
+For Kafka Connect you can get loggers information by accessing Connect API service:
+
+```
+$ kubectl exec -ti <kafka-cluster-name>-kafka-0 -- curl http://<connect-cluster-name>-connect-api:8083/admin/loggers
+```
+
+Since Kafka MirrorMaker2 is based on the Kafka Connect, getting its loggers is very similar:
+```
+$ kubectl exec -ti <kafka-cluster-name>-kafka-0 -- curl http://<mm2-cluster-name>-connect-api:8083/admin/loggers
```

Comment:
It is weird how you choose the pods to run the commands from. It might be cleaner to run it either directly from the pod where you execute it against localhost. Or run it from completely different pod which you start just for this command.

Reply:
Well, our docs says so: https://strimzi.io/docs/operators/master/using.html#property-kafka-connect-logging-reference
I have nothing against running commands directly from the (for example) connect pod. However, its name is variable so the example command is less copy-paste-able.

Reply:
Well, that is stupid from the docs then. And I do not think it makes it right here. The command is anyway not a direct copy-paste as it has the pod name placeholder. So you can equally well have the `<connect-pod-name>` placeholder.

Reply:
fixed